balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,620 @@
+	// present in only one endpoint before moving further.
+	newEndpoints = deDupAddresses(newEndpoints)
+
+	// Perform the optional shuffling described in gRFC A62. The shuffling will
+	// change the order of endpoints but not touch the order of the addresses
+	// within each endpoint. - A61
+	if cfg.ShuffleAddressList {
+		newEndpoints = append([]resolver.Endpoint{}, newEndpoints...)
+		internal.ShuffleAddressListForTesting.(func(int, func(int, int)))(len(newEndpoints), func(i, j int) { newEndpoints[i], newEndpoints[j] = newEndpoints[j], newEndpoints[i] })
+	}
+
+	// If the previous ready subconn exists in new address list,
+	// keep this connection and don't create new subconns.
+	prevAddr := b.addressList.currentAddress()
+	prevAddrsCount := b.addressList.size()
+	b.addressList.updateEndpointList(newEndpoints)
+	if b.state == connectivity.Ready && b.addressList.seekTo(prevAddr) {
+		return nil
+	}
+
+	b.reconcileSubConnsLocked(newEndpoints)
+	// If it's the first resolver update or the balancer was already READY
+	// (but the new address list does not contain the ready subconn) or
+	// CONNECTING, enter CONNECTING.
+	// We may be in TRANSIENT_FAILURE due to a previous empty address list,
+	// we should still enter CONNECTING because the sticky TF behaviour mentioned
+	// in A62 applies only when the TRANSIENT_FAILURE is reported due to connectivity
+	// failures.
+	if b.state == connectivity.Ready || b.state == connectivity.Connecting || prevAddrsCount == 0 {
+		// Start connection attempt at first address.
+		b.state = connectivity.Connecting
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Connecting,
+			Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+		})
+		b.requestConnectionLocked()
+	} else if b.state == connectivity.TransientFailure {
+		// If we're in TRANSIENT_FAILURE, we stay in TRANSIENT_FAILURE until
+		// we're READY. See A62.
+		b.requestConnectionLocked()
+	}
+	return nil
+}
+
+// UpdateSubConnState is unused as a StateListener is always registered when
+// creating SubConns.
+func (b *pickfirstBalancer) UpdateSubConnState(subConn balancer.SubConn, state balancer.SubConnState) {
+	b.logger.Errorf("UpdateSubConnState(%v, %+v) called unexpectedly", subConn, state)
+}
+
+func (b *pickfirstBalancer) Close() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.closeSubConnsLocked()
+	b.state = connectivity.Shutdown
+}
+
+// ExitIdle moves the balancer out of idle state. It can be called concurrently
+// by the idlePicker and clientConn so access to variables should be synchronized.
+func (b *pickfirstBalancer) ExitIdle() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.state == connectivity.Idle {
+		b.requestConnectionLocked()
+	}
+}
+
+func (b *pickfirstBalancer) closeSubConnsLocked() {
+	for _, sd := range b.subConns.Values() {
+		sd.(*scData).subConn.Shutdown()
+	}
+	b.subConns = resolver.NewAddressMap()
+}
+
+// deDupAddresses ensures that each address belongs to only one endpoint.
```

Comment:
According to Mark, this is an implicit requirement out of the resolver (I asked about WRR and Outlier Detection). What should the behavior be in this undefined case? For OD, Mark mentioned he just uses the first endpoint that an address maps too, but the scenario where an endpoint wraps the same address as another isn't really a valid scenario.

Reply:
I think that is in the context of EDS. Here, pick_first could be the top-level LB policy on the channel and there is no guarantee that the name resolver being used does not send duplicate addresses. Also, there is no complexity around localities here. The pick_first LB policy gets a flat list of endpoints or addresses.

Reply:
Yes, there is no guarantee, but Mark mentioned in the xDS Chat duplicate addresses is an implicit bug in the resolver emissions. So behavior in OD/WRR is undefined in that scenario (mainly deployed in xDS Tree where EDS has a requirement of no deduping).

Reply:
`pick_first` doesn't really care about endpoints, it will flatten the endpoint list and try to connect to all the addresses. The effect of for de-duplicating addresses is that we don't try to connect to the same address multiple times during a pass of pickfirst.
I've changed the implementation of `UpdateClientConnState` to flatten the endpoint list very early so that other parts of the LB policy handle a slice of addresses instead of a slice of endpoints. I initially started off converting the address list to endpoint list in `UpdateClientConnState` instead of the opposite after looking at Java's implementation, but its not really required.